Always close the `response_stream` after `HTTP.request`

nickrobinson251 opened the pull request:
- We need to explicitly close the `response_stream`
  (`HTTP.request` no longer closes it for us)
  in HTTP.jl v0.9.15+ (see https://github.com/JuliaWeb/HTTP.jl/pull/752)
- Since `close` is safe to call multiple times,
  this should be compatible with old HTTP.jl versions. 
- Should fix issue #466 (and https://github.com/JuliaCloud/AWSS3.jl/issues/215)
- Also since we're fixing this here, we can close https://github.com/JuliaWeb/HTTP.jl/issues/772

~TODO: add test... i just need to recreate the issue using AWS.jl explicitly (rather than AWSS3)~ done.

fredrikekre commented on src/utilities/request.jl:
```diff
@@ -213,6 +213,8 @@ function _http_request(http_backend::HTTPBackend, request::Request)
                         (isa(e, HTTP.StatusError) && _http_status(e) >= 500)
         end
+    finally
+        close(request.response_stream)
```
Probably should be conditional on if a BufferStream was actually initialized on L193? I don't know this code at all though, but seems like `.response_stream` is at least sometimes `nothing`?

nickrobinson251 replied:
good point!

```suggestion
        request.response_stream === nothing || close(request.response_stream)
```

oxinabox replied:
Maybe:
```suggestion
        request.response_stream isa IO && close(request.response_stream)
```

ericphanson commented:
I'm a bit worried about the `finally`, since I see no other uses in this package where we use Retry.jl's macros with `finally`, and there are no tests for it: https://github.com/JuliaWeb/Retry.jl/blob/master/test/runtests.jl.

nickrobinson251 commented:
> I'm a bit worried about the `finally`, since I see no other uses in this package where we use Retry.jl's macros with `finally`

i suppose we could just call `close` in both branches of the `try`, if we want to avoid using `finally`?

e.g.
```diff
-        return @mock HTTP.request(
+        resp = @mock HTTP.request(
             http_stack,
             request.request_method,
             HTTP.URI(request.url),
             HTTP.mkheaders(request.headers),
             request.content;
             require_ssl_verification=false,
             response_stream=request.response_stream,
             http_options...,
         )
+        request.response_stream === nothing || close(request.response_stream)
+        return resp
     catch e
+        request.response_stream === nothing || close(request.response_stream)
         # Base.IOError is needed because HTTP.jl can often have errors that aren't
```

ericphanson commented:
The semantics of these macros are pretty confusing; both the `try` and `catch` branches can be executed multiple times if it retries, and we don't want to close the stream if we're going to retry. So I kind of think we need `finally`... just need to be sure it actually works 😅.

ericphanson commented:
Let’s see what CI says already— we can check at least if tests pass here with the new HTTP.jl release.

bors try

ericphanson commented:
Tests pass and we are using HTTP v0.9.15: https://github.com/JuliaCloud/AWS.jl/runs/3733253932#step:10:48

so that’s promising

nickrobinson251 commented:
What's a sensible way to test something that causes code to hang?

For now i've just put some code (which hangs on `master`) into a test, but i'm not sure if that's the best way to go (since we rely on CI timing out or, if locally, realising something is wrong): https://github.com/JuliaCloud/AWS.jl/pull/467/commits/4dcecc6686537b563b953010fbee2b2e16efc07a

ericphanson commented:
We at least have timeouts now (https://github.com/JuliaCloud/AWS.jl/blob/master/.github/workflows/CI.yml#L16) due to hangs in the past, so I think the way you did it is OK.

Should we be creating buckets, or is there one we should use already @mattBrzezinski ?

nickrobinson251 commented:
> Should we be creating buckets

ah yeah, i meant to ask -- the `testset`s in `test/issues.jl` all set up and then teardown buckets, which is kinda nice but also kinda slow and make for a lot of boilerplate. For now i've just followed the same pattern, but (perhaps as a follow-up) maybe all these `issues.jl` tests on S3 could just use different files in the same bucket?

mattBrzezinski commented on src/utilities/request.jl:
```diff
@@ -213,6 +213,8 @@ function _http_request(http_backend::HTTPBackend, request::Request)
                         (isa(e, HTTP.StatusError) && _http_status(e) >= 500)
         end
+    finally
+        request.response_stream isa IO && close(request.response_stream)
```
If an end-user were to pass in their own `response_stream`, what would happen here if it closes? Is that data still available in it, or does it go away?

fredrikekre replied:
I was gonna suggest something like
```diff
diff --git a/src/utilities/request.jl b/src/utilities/request.jl
index 12debd4..2ef8363 100644
--- a/src/utilities/request.jl
+++ b/src/utilities/request.jl
@@ -189,11 +189,13 @@ function _http_request(http_backend::HTTPBackend, request::Request)
     @repeat 4 try
         http_stack = HTTP.stack(; redirect=false, retry=false, aws_authorization=false)
 
+        should_close = false
         if request.return_stream && request.response_stream === nothing
             request.response_stream = Base.BufferStream()
+            should_close = true
         end
 
-        return @mock HTTP.request(
+        r = @mock HTTP.request(
             http_stack,
             request.request_method,
             HTTP.URI(request.url),
@@ -203,6 +205,10 @@ function _http_request(http_backend::HTTPBackend, request::Request)
             response_stream=request.response_stream,
             http_options...,
         )
+        if should_close
+            close(request.response_stream)
+        end
+        return r
     catch e
         # Base.IOError is needed because HTTP.jl can often have errors that aren't
         # caught and wrapped in an HTTP.IOError
```
earlier. E.g. only close if it was opened internally.

nickrobinson251 replied:
the behaviour is exactly as before (i.e. as we had with older HTTP.jl versions). In older HTTP versions, `close` was already called inside the `HTTP.request` above ﻿https://github.com/JuliaWeb/HTTP.jl/pull/752/files#diff-2877f0a59db9c7dda95204097d3c2010021eb8f6c59c98d56d3f7d16e38b0acbL149

ericphanson replied:
> the behaviour is exactly as before (i.e. as we had with older HTTP.jl versions)

I think the idea is that this maybe wasn't correct from HTTP.jl which is why it got changed there. https://github.com/JuliaCloud/AWS.jl/pull/396#discussion_r694661438 and #433  have lots of discussion on this (which I forgot completely about until just now)

nickrobinson251 replied:
> only close if it was opened internally

wouldn't this be different behaviour to what we had when using older HTTP.jl versions? 
i.e. didn't `HTTP.request` just call `close` on the `response_stream` (if it wasn't `nothing`)?

(i don't know, since i'm not too familiar with this stuff)

fredrikekre replied:
Yea probably. I guess the motivation for the change in HTTP.jl is that people would like to handle the closing themselves, and maybe the same argument can apply here too. I wasn't very involved in those discussions though.

mattBrzezinski replied:
I have not been paying too much attention to the HTTP.jl changes. Are there use cases anyone can think of to keeping the streams open and having users manually close them?

Personally I think closing the HTTP requests automatically makes sense. It gives a loose contract of, this is the data which AWS has provided you from your request.

Maybe I'm missing something obvious though.

fredrikekre replied:
Yea, maybe it should be reverted.

nickrobinson251 replied:
FWIW, i have no opinion on how best to fix this. But i am keen to fix it as fast as possible, as not being able get files from S3 is quite debilitating for some workflows at Invenia 😊

oxinabox replied:
Yes, I would say we should merge this PR which basically brings back the behavior of HTTP 0.9.14 and then continue this discussion in an issue.

I would really like to see this merged today.

nickrobinson251 replied:
https://github.com/JuliaCloud/AWS.jl/issues/468

mattBrzezinski commented on test/issues.jl:
```diff
@@ -81,3 +81,24 @@ end
 end
+
+@testset "issue 466" begin
+    bucket_name = "aws-jl-test-issues---" * _now_formatted()
```
Do we need to create a new bucket for this test? This file creates one at the beginning.

nickrobinson251 replied:
the testsets above need a little bit of re-jigging if we're to use that same bucket throughout -- i've made that change as i think it make most sense to just use that one bucket when possible

mattBrzezinski reviewed (commented):
LGTM, left a couple comments.

mattBrzezinski commented on test/issues.jl:
```diff
@@ -1,21 +1,20 @@
-bucket_name = "aws-jl-test-issues---" * _now_formatted()
-S3.create_bucket(bucket_name)
+const BUCKET_NAME = "aws-jl-test-issues---" * _now_formatted()
+S3.create_bucket(BUCKET_NAME)
```
This might need a bit better of a solution, but we should wrap all these `@testset` in a `try/finally` and delete the bucket afterwards. Otherwise on any failure it'll continue to exist.

nickrobinson251 replied:
done https://github.com/JuliaCloud/AWS.jl/pull/467/commits/27dfaee697cd1d1f223cbe9f3c98657fdfc9d8c9

mattBrzezinski reviewed (approved):
One comment about the `try/finally` in `test/issues.jl` and it LGTM!

mattBrzezinski commented:
FWIW You should be able to just run JuliaFormatter w/ BlueStyle on the entire package and it'll make the CI step happy. I believe I already cleaned this repo out.

ericphanson commented:
I'm concerned about this change; I think https://github.com/JuliaCloud/AWS.jl/pull/467#discussion_r717666189 is probably the right way. 

In https://github.com/JuliaCloud/AWS.jl/pull/396#discussion_r695321404 @vtjnash said
> Whoever created the stream is generally responsible for marking when it is done, or passing it to another function which will take care of that.

which sounds like we should only be closing the ones we create here.

mattBrzezinski commented:
> I'm concerned about this change; I think [#467 (comment)](https://github.com/JuliaCloud/AWS.jl/pull/467#discussion_r717666189) is probably the right way.
> 
> In [#396 (comment)](https://github.com/JuliaCloud/AWS.jl/pull/396#discussion_r695321404) @vtjnash said
> 
> > Whoever created the stream is generally responsible for marking when it is done, or passing it to another function which will take care of that.
> 
> which sounds like we should only be closing the ones we create here.

I haven't had time to read through all these threads, but will soon(TM). I think we can push these changes through as they mimic the existing functionality, then open an issue to discuss?

ericphanson commented:
Ok, that sounds fine to me

mattBrzezinski commented:
bors r+

mattBrzezinski commented:
CI failed, but also threw

> WARNING: redefinition of constant BUCKET_NAME. This may fail, cause incorrect answers, or produce other errors.

nickrobinson251 commented on test/issues.jl:
```diff
@@ -1,83 +1,102 @@
-    finally
-        S3.delete_object(bucket_name, file_name)
-        S3.delete_bucket(bucket_name)
+const BUCKET_NAME = "aws-jl-test-issues---" * _now_formatted()
```
this file get's `include`d multiple times

```suggestion
BUCKET_NAME = "aws-jl-test-issues---" * _now_formatted()
```

nickrobinson251 commented:
i don't understand bors -- is CI running, or...?

mattBrzezinski commented:
bors r+

omus commented:
Personally I think the way this should have been address was to upper bound HTTP.jl. That would have been a trivial change in comparison and would have given us more time to work out the details with this PR

oxinabox commented:
A fair point conceptually.
But in practice an incredibly annoying one, do to how julia's resolver/registry works.
since would have needed to go and retroactively capped the version of HTTP.jl in all past releases of  AWS.jl.
Which would have required manual registry edits.
(I have suggested before that for post 1.0 packages, like AWS.jl the registry should refuse to install any but the lasted patch release which would have fixed it.)

Another good option might have been to revert the change to HTTP.jl  and tag a release.
(since it was supposedly a nonbreaking change but did infact break a pretty core package for a lot of people) 
While we worked out weather or not it was breaking after all, or wether AWS was violating AWS.jl's public API (and if so fixed AWS.jl)

nickrobinson251 commented:
Yeah, reverting the change to HTTP.jl would have been another option, but may have required more discussion.
Pinning HTTP is quite painful, as Lyndon says. This PR was a 2 line change to the `src` (the big diffs to the tests were various requests made in code review, mostly addressing issues in the existing tests, arguably out of scope for this PR tbh). And since this just maintains the existing (pre-HTTP.jl v0.9.15) behaviour in an internal function, i don't really see the harm. As far as this package is concerned, it just protects against the relevant change in HTTP. Especially as a change like #468 can be done, and maybe should be done, (just like it could have been before).
